extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -121,22 +118,26 @@ int fill(final long position, final ByteBuffer dest) throws IOException {
             final int impliedReadAhead = (int) (lastFragmentIx - firstFragmentIx);
             final int desiredReadAhead = instructions.readAheadCount();
             final long totalRemainingFragments = numFragments - firstFragmentIx - 1;
-            final int maxReadAhead = instructions.maxCacheSize() - 1;
-            readAhead = Math.min(
-                    Math.max(impliedReadAhead, desiredReadAhead),
-                    (int) Math.min(maxReadAhead, totalRemainingFragments));
+            readAhead = Math.min(Math.max(impliedReadAhead, desiredReadAhead), totalRemainingFragments);
         }
-        final S3Request firstRequest = getOrCreateRequest(firstFragmentIx);
+        // Hold a reference to the first request to ensure it is not evicted from the cache
+        acquiredRequest = getOrCreateRequest(firstFragmentIx);
         for (int i = 0; i < readAhead; ++i) {
+            // Do not hold references to the read-ahead requests
             getOrCreateRequest(firstFragmentIx + i + 1);
         }
         // blocking
-        int filled = firstRequest.fill(position, dest);
+        int filled = acquiredRequest.request.fill(position, dest);
+        acquiredRequest = null;
+
         for (int i = 0; dest.hasRemaining(); ++i) {
-            // Since we have already created requests for read ahead fragments, we can retrieve them from the local
-            // cache
-            final S3Request request = getRequestFromLocalCache(firstFragmentIx + i + 1);
-            if (request == null || !request.isDone()) {
+            final S3Request.AcquiredRequest acquiredReadAheadRequest =
+                    getRequestFromSharedCache(firstFragmentIx + i + 1);
+            if (acquiredReadAheadRequest == null) {
+                break;
+            }
+            final S3Request request = acquiredReadAheadRequest.request;
+            if (!request.isDone()) {
```

Comment:
This is also making me look into the fill logic. I wonder if fill should instead take in `AcquiredRequest` and just use `ownershipToken` instead of `get()` (and / or, use `get()` as a safety check to ensure it's the same object as `ownershipToken`). Or change the API a bit more, and move `fill` into `AcquiredRequest`, I think that makes structural sense?

Reply:
If fill iis moved into AcquiredRead, I think safety check on get() is unneccessary since tryAcquire is pretty straightforward.

Reply:
Replied here: https://github.com/deephaven/deephaven-core/pull/5613#discussion_r1640014443